src/soc/ibm/power9/fsi.h

```diff
@@ -0,0 +1,51 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#ifndef __SOC_IBM_POWER9_FSI_H
+#define __SOC_IBM_POWER9_FSI_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+void fsi_init(void);
+
+/* This isn't meant to be used directly */
+uint32_t fsi_op(uint8_t chip, uint32_t addr, uint32_t data, bool is_read);
+
+/* FSI-functions operate on byte addresses */
+
+static inline uint32_t read_fsi(uint8_t chip, uint32_t addr)
+{
+	return fsi_op(chip, addr, /*data=*/0, /*is_read=*/true);
+}
+
+static inline void write_fsi(uint8_t chip, uint32_t addr, uint32_t data)
+{
+	(void)fsi_op(chip, addr, data, /*is_read=*/false);
+}
+
+/* CFAM-functions are FSI-functions that operate on 4-byte word addresses */
+
+static inline uint32_t cfam_addr_to_fsi(uint32_t cfam)
+{
+	const uint32_t CFAM_ADDRESS_MASK = 0x1FF;
+	const uint32_t CFAM_ENGINE_OFFSET_MASK = 0xFE00;
```

Comment:
With such masks, address after multiplication can potentially overlap with engine offset part. I know this is the same as in Hostboot, but maybe leave a comment mentioning we are aware of it, because I doubt we can find any documentation that would tell which one of those fields should be shrunk.